### PR TITLE
Caching toots - attempt 2

### PR DESCRIPTION
**Pull request by charlag**
I've tried to rebase #663 but it was branched from another branch and it is too old so I failed.
Instead I've `cherry-pick`ed last 5 commits from that branch which seem to contain useful changes.
I did not try it yet because I'm out of energy already but it can be useful for someone to inspect or fix.

### CODE REVIEW

**Comment by charlag** [comment]
I've updated PR to fix some errors. At least some caching is working now but I am still not sure how to mark gaps in DB

**Comment by connyduck** [comment]
The offline part works, very well done.
Three major bugs immediately caught my eye: Can't scroll timeline more than one page anymore, database migration is missing & multiple accounts get their timelines mixed up.
Why can't you save a almost empty status (everything null) as placeholder?

**Comment by charlag** [comment]
Thanks for checking it out!

1. It was there but apparently broke, didn't test it now
2. It was there but got lost, again
3. Because I sort statuses by ID. Which ID should I insert for bogus status? Can I sort them by time? Who knows

**Comment by connyduck** [comment]
> Because I sort statuses by ID. Which ID should I insert for bogus status? Can I sort them by time? Who knows

what if when a placeholder is encountered, we only save the upper part of the timeline?

**Comment by charlag** [comment]
And delete everything below? That would be unpleasant

**Comment by connyduck** [comment]
Yes, but it should be a cache and not a save-the-whole-timeline-forever thing, shouldn't it?

Another possible solution could be to use one of the ids that presumably belong inside the gap to the placeholder? The id of the toot above the gap minus 1 or something?

**Comment by charlag** [comment]
Uuugh, maybe my problem is that I was trying to do exactly this? With future search and everything.

You proposal could lead to clashes easily, especially on small instances and I would rather not.
Maybe we need a flag or something like that? Like "gap below"

**Comment by connyduck** [comment]
It definitely needs some clean up logic to prevent database size to grow forever

**Comment by charlag** [comment]
Yup, but that's another problem and it's pretty easy to solve imo

**Comment by charlag** [comment]
>multiple accounts get their timelines mixed up

not sure why it could happen because `WHERE s.timelineUserId = :account `
:thinking:

**Comment by charlag** [comment]
I've tried adding a migration but didn't really check it yet.
I've pretty much just copied it from the scheme and checked in the scheme itself because... it's a good practice and makes your life easier I guess? Can remove it if there will be objections but I think it's not such a bad idea (bonus: can test migrations with in-memory DB. At least I think so)

**Comment by charlag** [comment]
Some progress on this
I've changed caching strategy. Now initial load is DB-only, after that data is fetched from the server to show updated content.
For all other requests it fetches from network but falls back to disk. Maybe there should be a better way than catching IOExceptions. It doesn't fix "gap problem" but makes it "offline only" which is acceptable IMO and it seems like Pinafore works like that too.

I've also fixed some problems with duplicated DB entries. Mixing statuses between accounts should be fixed now but I didn't check. Need to save some more account fields to work correctly.

There's a hack for the initial request which asks for statuses with margin of 11 which seems to be working but maybe not always and we should figure this out.

We can improve thread loading by not reloading the toot itself now when we have it in the DB.

**Comment by charlag** [comment]
Also not forget to create an index on timeline account entities as Room proposes, otherwise updates will be slow

Also don't forget to update statuses on events

**Comment by charlag** [comment]
Rebased on master, fixed couple of small things, will check how it'll work.
(actually works kinda weird because on pull we scroll to top)
Now I'm trying to get current statuses but I keep newer statuses too.
I'm kinda lost, not sure what's missing now. Would appreciate some feedback but no pressure.

**Comment by charlag** [comment]
OK, first pull-to-refresh doesn't work, uuugh

**Comment by charlag** [comment]
Works better now, maybe we can really get there.
I think that we should do two requests on startup now: one for refreshing cached list and another one to get the latest things (to insert them on top). Will try to try that tomorrow

**Comment by charlag** [comment]
Hey, I know I've said that you can review it but I already see some conceptual problems.
Users want to retain their position but we don't do that and we can't actually do that until we have some way to mark (damn) gaps. Maybe idea with empty status is not so bad and we can do that. Will try to describe today but it may be tricky still.

**Comment by connyduck** [comment]
I tested a build from commit 7f0836e. Here is what I noticed:

- Migrations are not working
- After deleting data everything is mostly fine
- when killing app and restarting, the timeline is jumping in a weird way (I can make a video if you want)
- when pulling to refresh and there are new toots loading, timeline scrolls up way more than previously

**Comment by charlag** [comment]
Thx for testing! I think most of it should be solved by the next commit now but we will need to re-test anyway

**Comment by charlag** [comment]
Uff, much stuff today.
I think most things are in place. Biggest problems are:
- Sometimes there are two placeholders next to each other and they are broken in this case
- Almost no tests

**Comment by charlag** [comment]
One of the biggest remainig problems is "too many placeholders" (thing are not so bad if it's a problem I think).
While for toots from DB it is acceptable (but we still may avoid it) the most annoying one is one on the top which we get every time for some reason. 
Once we solve this (and migration) we should be done.

We should discuss if we want to include saving position in this PR or implement it later. It shouldn't be a big problem but this is already huge.

**Comment by charlag** [comment]
I've spent quite some [time](https://gist.github.com/charlag/3d13c6b0df85286f9ce5d63c95ad10e7) trying to find out why timeline Mastodon API is acting so weirdly.
I'm not sure yet if [this](
https://github.com/tootsuite/mastodon/pull/8736) will help but seems so.
Until it's widespread we will have do deal with some extra placeholders (or spam the server but it's harder).

@connyduck can I have your opinion on including position remembering in this PR or not?

**Comment by connyduck** [comment]
Installing new build right now, will give you feedback tomorrow!

Lets leave remembering position for now and focus on getting this merged, we can add it later. I think it should not be too much effort then.

**Comment by charlag** [comment]
Thanks! I won't be near PC for the weekend but next week I can apply feedback! 
Yeah, it's not going to be much.
Maybe we should also add cleanup (something like deleting all toots older than two weeks or something? Maybe except own)

**Comment by connyduck** [comment]
It seems that migrations work now!
But ... the home timeline does not load anymore, it stays on the progress bar forever ._.

**Comment by charlag** [comment]
Oh, it's not like that for me, probably I forgot to test with empty timeline... or... I don't even know.
I'll try next week if you won't find it first, maybe it's obvious

**Comment by connyduck** [comment]
Ok restarting the app or changing account gets the timeline shown again. But now there is a "load more" button on the top of the timeline and some seemingly random ones inbetween statuses.

**Comment by charlag** [comment]
It shouldn't generally appear on top and im between it should only appear for cached things 🤔

**Comment by charlag** [comment]
I think I've screwed up cleanup somehow...

**Comment by charlag** [comment]
Cleanup was embarrassingly broken (it deleted recent cached toots by everyone but you and you've mostly seen your toots... lol) but now it should be good, I've fixed test too.

**Comment by charlag** [comment]
I've tried to reproduce "mixed timeline" issue with multiple accounts on the same instance but it doesn't seem to happen to me.

I've found one minor problem on initial loading. The first toot in timeline disappears when we try to do initial update (thanks Mastodon API) and then only reappears when we load above. This makes it look like the whole timeline shifted. We can only really fix that with min_id API methods I think but I will try to mitigate it.

**Comment by charlag** [comment]
I've made a workaround for the issue.
I think that mixed timelines you've experienced were caused by the old DB from one of the previous versions but I'm not sure. Would be glad if you could confirm it (sometime later maybe)

File: app/src/main/java/com/keylesspalace/tusky/entity/Status.kt [hunk 1]
```diff
@@ -21,16 +21,16 @@ import java.util.*
...
         @SerializedName("in_reply_to_id") var inReplyToId: String?,
         @SerializedName("in_reply_to_account_id") val inReplyToAccountId: String?,
         val reblog: Status?,
         val content: Spanned,
         @SerializedName("created_at") val createdAt: Date,
         val emojis: List<Emoji>,
-        @SerializedName("reblogs_count") val reblogsCount: Int,
-        @SerializedName("favourites_count") val favouritesCount: Int,
+        @SerializedName("reblogs_count") val reblogsCount: Long,
+        @SerializedName("favourites_count") val favouritesCount: Long,
```
[the remark this answers is not in the document]
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/entity/Status.kt]: Why 65k? `Integer.MAX_VALUE` is 2 billion and something

**Comment by connyduck** [comment]
Ok, weird timeline jumping seems to be gone, but wrong timeline is still here.
But I have a better bug description: It has nothing to do with being on the same instance.
Steps to reproduce: 
- open app
- see timeline being shown correctly
- add another account
- **notice timeline is the same as for the previous account**
- switch accounts: all timelines are shown correctly again

additionally: I have an account with an empty timeline, it just shows the progressbar forever

**Comment by charlag** [comment]
Sorry, I cannot reproduce either!
I think but with "not my timeline" appeared once when I was logging out but I cannot make it happen again :c
My guess was that account was switched later than statuses were requested but it doesn't happen anymore!
I have to ask you to try to nail it down when possible :bowing_man:

**Comment by charlag** [comment]
(also I'm not sure if you had clean DB when testing)

File: app/src/main/java/com/keylesspalace/tusky/entity/Status.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/entity/Status.kt]: Yes, you're right, I thought it's platform dependent for some reason. My bad. I can change it back.

**Comment by connyduck** [comment]
Alright here is a video: https://connyduck.at/tusky_1.mp4
My mastodon.sergal.org account follows nobody and has an empty timeline.

**Comment by charlag** [comment]
Hey, thanks for the video, I believe you ofc!
Maybe I should test some emtpy account.

**Comment by charlag** [comment]
Thanks for the feedback :bowing_man: 
I'll try to work on it in Janruary

File: app/src/main/java/com/keylesspalace/tusky/repository/TimelineRepository.kt [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+                followingCount = 0,
+                followersCount = 0,
+                statusesCount = 0,
+                source = null,
+                // TODO actually fill these fields
+                bot = false,
+                emojis = null,
+                fields = null,
+                moved = null
+        )
```
[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/repository/TimelineRepository.kt]: Then I need JVMoverloads... it's basically the only place it's created so I'm hesitant to change it tbh

File: app/src/main/java/com/keylesspalace/tusky/repository/TimelineRepository.kt [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+                note = SpannedString(""),
+                url = url,
+                avatar = avatar,
+                header = "",
+                locked = false,
+                followingCount = 0,
+                followersCount = 0,
+                statusesCount = 0,
+                source = null,
+                // TODO actually fill these fields
```
[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/repository/TimelineRepository.kt]: Should we leave it for offline or it's out of scope and I should just add emojis?

[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/repository/TimelineRepository.kt]: Wrong question: should I add emoji only? I think I will do that

[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/repository/TimelineRepository.kt]: Oh, I thought emojis in display names are not allowed anymore for accessibility reason or something else?

**Comment by charlag** [comment]
Found a very stupid bug:

1. Open client
2. Disable network
3. Scroll to bottom till the end
4. Have "Load more" button at the bottom (?!)
5. Press it and crash :cry: 

It's not hard to handle this case but point is that there should be no "load more" button there. I'll try to find out why it's there.

**Comment by charlag** [comment]
I fixed some bugs but it doesn't look terribly solid overall. I will try to test it more in the next few days with different network conditions.

**Comment by charlag** [comment]
😱

**Pull request merged**